include/oneapi/dpl/pstl/hetero/histogram_impl_hetero.h

```diff
@@ -0,0 +1,174 @@
+    return __binhash_manager_custom_boundary{::std::move(__bin_hash_range), ::std::move(__buffer_lifetime_holder)};
+}
+
+template <typename _Name>
+struct __hist_fill_zeros_wrapper
+{
+};
+
+template <typename _ExecutionPolicy, typename _RandomAccessIterator1, typename _Size, typename _BinHash,
+          typename _RandomAccessIterator2>
+oneapi::dpl::__internal::__enable_if_hetero_execution_policy<_ExecutionPolicy>
+__pattern_histogram(_ExecutionPolicy&& __exec, _RandomAccessIterator1 __first, _RandomAccessIterator1 __last,
+                    _Size __num_bins, _BinHash&& __func, _RandomAccessIterator2 __histogram_first)
+{
+    //If there are no histogram bins there is nothing to do
+    if (__num_bins > 0)
+    {
+        using _global_histogram_type = typename ::std::iterator_traits<_RandomAccessIterator2>::value_type;
+        const auto __n = __last - __first;
+
+        // The access mode we we want here is "read_write" + no_init property to cover the reads required by the main
+        //  kernel, but also to avoid copying the data in unnecessarily.  In practice, this "write" access mode should
+        //  accomplish this as write implies read, and we avoid a copy-in from the host for "write" access mode.
+        // TODO: Add no_init property to get_sycl_range to allow expressivity we need here.
+        auto __keep_bins =
+            oneapi::dpl::__ranges::__get_sycl_range<oneapi::dpl::__par_backend_hetero::access_mode::write,
+                                                    _RandomAccessIterator2>();
+        auto __bins_buf = __keep_bins(__histogram_first, __histogram_first + __num_bins);
+        auto __bins = __bins_buf.all_view();
+
+        auto __fill_func = oneapi::dpl::__internal::fill_functor<_global_histogram_type>{_global_histogram_type{0}};
+        //fill histogram bins with zeros
+
+        auto __init_event = oneapi::dpl::__par_backend_hetero::__parallel_for(
+            oneapi::dpl::__par_backend_hetero::make_wrapped_policy<__hist_fill_zeros_wrapper>(__exec),
+            unseq_backend::walk_n<_ExecutionPolicy, decltype(__fill_func)>{__fill_func}, __num_bins, __bins);
+
+        if (__n > 0)
+        {
+            //need __binhash_manager to stay in scope until the kernel completes to keep the buffer alive
+            // __make_binhash_manager will call __get_sycl_range for any data which requires it within __func
+            auto __binhash_manager = __make_binhash_manager(::std::forward<_BinHash>(__func));
+            auto __keep_input =
+                oneapi::dpl::__ranges::__get_sycl_range<oneapi::dpl::__par_backend_hetero::access_mode::read,
+                                                        _RandomAccessIterator1>();
+            auto __input_buf = __keep_input(__first, __last);
+
+            __parallel_histogram(::std::forward<_ExecutionPolicy>(__exec), __init_event, __input_buf.all_view(),
+                                 ::std::move(__bins), ::std::move(__binhash_manager))
```

Comment:
Are you sure that you want moving `__binhash_manager`?
The fact is it contains `__buffer_lifetime_holder`. M.b. we should keep the lifetime in this scope? 
I am asking to be sure that we don't lose "control" here.